File: python/src/dexcost/auto_task.py
"""Auto-task creation for auto-instrumented calls without an explicit task."""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from decimal import Decimal

# Re-export set_context/get_context for backward compat with tests
from dexcost.context import (
    DexcostContext as DexcostContext,
)
from dexcost.context import (
    clear_context as clear_context,
)
from dexcost.context import (
    get_context,
    get_current_task,
)
from dexcost.context import (
    set_context as set_context,
)
from dexcost.models.event import Event
from dexcost.models.task import Task

# ---------------------------------------------------------------------------
# Public helpers
# ---------------------------------------------------------------------------


def needs_auto_task() -> bool:
    """Return True if there is no active explicit task."""
    return get_current_task() is None


def create_auto_task(task_type: str) -> Task:
    """Create a task with attribution from the current DexcostContext.

    Agent and workflow identity remain separate from *task_type*. When the
    context contains business identity, the automatic task becomes its own
    canonical root so the identity snapshot is deliverable.
    """
    ctx = get_context()
    task_id = uuid.uuid4()
    has_business_identity = ctx is not None and any(
        (
            ctx.customer_id,
            ctx.project_id,
            ctx.user_id,
            ctx.product_id,
            ctx.agent,
            ctx.workflow_id,
        )
    )
    return Task(
        task_id=task_id,
        task_type=task_type,
        status="pending",
        started_at=datetime.now(timezone.utc),
        customer_id=ctx.customer_id if ctx else None,
        project_id=ctx.project_id if ctx else None,
        user_id=ctx.user_id if ctx else None,
        product_id=ctx.product_id if ctx else None,
        root_task_id=task_id if has_business_identity else None,
        agent_id=ctx.agent if ctx else None,
        agent_version=ctx.agent_version if ctx else None,
        workflow_id=ctx.workflow_id if ctx else None,
        workflow_session_id=ctx.workflow_session_id if ctx else None,
        metadata=dict(ctx.metadata) if ctx and ctx.metadata else {},
    )


def finalize_auto_task(task: Task, event: Event, status: str = "success") -> None:
    """Finalize an auto-task: aggregate the event's cost and set end time."""
    task.status = status
    task.ended_at = datetime.now(timezone.utc)

    cost = event.cost_usd
    if event.event_type == "llm_call":
        task.llm_cost_usd = cost
        task.total_input_tokens = event.input_tokens or 0
        task.total_output_tokens = event.output_tokens or 0
        task.total_cached_tokens = event.cached_tokens or 0
    elif event.event_type == "external_cost":
        task.external_cost_usd = cost
    elif event.event_type == "compute_cost":
        task.compute_cost_usd = cost
    elif event.event_type == "retry_marker":
        task.retry_count = 1
        task.retry_cost_usd = cost

    task.total_cost_usd = cost

    if event.is_retry:
        task.retry_count = (task.retry_count or 0) + 1
        task.retry_cost_usd = (task.retry_cost_usd or Decimal(0)) + cost

    if status == "failed":
        task.failure_count = 1

File: rust/src/core/auto_task.rs
//! Auto-task creation and finalization for instrumented calls without explicit
//! task context.
//!
//! When a wrapper client records an LLM call and no explicit task is active,
//! an auto-task is created to group the cost event. Once the call completes,
//! the auto-task is finalized with aggregated costs.

use chrono::Utc;

use crate::core::context::get_dexcost_context;
use crate::core::models::{CostEvent, EventType, Task, TaskStatus};
use crate::transport::buffer::EventBuffer;

/// Creates an auto-task for instrumented calls without explicit task context.
///
/// Reads `customer_id` and `project_id` from the ambient
/// [`DexcostContext`](crate::DexcostContext) if available. Sets metadata
/// indicating this is a session-initiated auto-task.
///
/// # Arguments
///
/// * `task_type` - The type/name for this auto-task (e.g. `"openai.chat"`).
pub async fn create_auto_task(task_type: &str) -> Task {
    let ctx = get_dexcost_context().await;
    let mut task = Task::new(task_type);

    if let Some(c) = ctx {
        task.customer_id = c.customer_id;
        task.project_id = c.project_id;
    }

    task.metadata
        .insert("session".to_string(), serde_json::Value::Bool(true));
    task.metadata.insert(
        "initiated_by".to_string(),
        serde_json::Value::String(task_type.to_string()),
    );

    task
}

/// Finalizes an auto-task: aggregates event costs, sets status, persists to buffer.
///
/// If an event is provided, its cost and token fields are aggregated into the
/// task totals. Retry metrics are updated when the event has `is_retry` set.
///
/// # Arguments
///
/// * `task` - Mutable reference to the task being finalized.
/// * `event` - Optional cost event to aggregate into the task.
/// * `status` - The final status string (`"success"` or `"failed"`).
/// * `buffer` - The event buffer used to persist the updated task.
pub fn finalize_auto_task(
    task: &mut Task,
    event: Option<&CostEvent>,
    status: &str,
    buffer: &mut EventBuffer,
) {
    task.status = match status {
        "success" => TaskStatus::Success,
        "failed" => TaskStatus::Failed,
        _ => TaskStatus::Pending,
    };
    task.ended_at = Some(Utc::now());

    if let Some(ev) = event {
        match ev.event_type {
            EventType::LlmCall => task.llm_cost_usd += ev.cost_usd,
            EventType::ExternalCost => task.external_cost_usd += ev.cost_usd,
            EventType::ComputeCost => task.compute_cost_usd += ev.cost_usd,
            EventType::RetryMarker => { /* retry markers don't add to category costs */ }
        }
        task.total_cost_usd = task.llm_cost_usd + task.external_cost_usd + task.compute_cost_usd;

        if let Some(tokens) = ev.input_tokens {
            task.total_input_tokens += tokens;
        }
        if let Some(tokens) = ev.output_tokens {
            task.total_output_tokens += tokens;
        }
        if let Some(tokens) = ev.cached_tokens {
            task.total_cached_tokens += tokens;
        }
        if ev.is_retry {
            task.retry_count += 1;
            task.retry_cost_usd += ev.cost_usd;
        }
    }

    buffer.upsert_task(task.clone());
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::core::context::{clear_dexcost_context, set_context, DexcostContext};
    use crate::core::models::{CostEvent, EventType, TaskStatus};
    use crate::transport::buffer::EventBuffer;
    use rust_decimal::Decimal;

    // Process-wide serialisation lock — the global context is shared.
    use std::sync::LazyLock;
    use tokio::sync::Mutex;

    static CTX_LOCK: LazyLock<Mutex<()>> = LazyLock::new(|| Mutex::new(()));

    #[tokio::test]
    async fn test_create_auto_task_with_context() {
        let _g = CTX_LOCK.lock().await;

        set_context(DexcostContext {
            customer_id: Some("auto-cust".into()),
            project_id: Some("auto-proj".into()),
            metadata: None,
            agent: None,
        })
        .await;

        let task = create_auto_task("openai.chat").await;
        assert_eq!(task.task_type, "openai.chat");
        assert_eq!(task.customer_id.as_deref(), Some("auto-cust"));
        assert_eq!(task.project_id.as_deref(), Some("auto-proj"));
        assert_eq!(task.status, TaskStatus::Pending);
        assert_eq!(
            task.metadata.get("session"),
            Some(&serde_json::Value::Bool(true))
        );
        assert_eq!(
            task.metadata.get("initiated_by"),
            Some(&serde_json::Value::String("openai.chat".to_string()))
        );

        clear_dexcost_context().await;
    }

    #[tokio::test]
    async fn test_create_auto_task_without_context() {
        let _g = CTX_LOCK.lock().await;

        clear_dexcost_context().await;
        let task = create_auto_task("test.call").await;
        assert!(task.customer_id.is_none());
        assert!(task.project_id.is_none());
        assert_eq!(task.task_type, "test.call");
    }

    #[tokio::test]
    async fn test_finalize_auto_task_with_llm_event() {
        let _g = CTX_LOCK.lock().await;
        clear_dexcost_context().await;

        let mut task = create_auto_task("openai.chat").await;
        let mut event = CostEvent::new(&task.task_id, EventType::LlmCall);
        event.cost_usd = Decimal::new(5, 2); // 0.05
        event.input_tokens = Some(1000);
        event.output_tokens = Some(500);

        let mut buffer = EventBuffer::new().unwrap();
        finalize_auto_task(&mut task, Some(&event), "success", &mut buffer);

        assert_eq!(task.status, TaskStatus::Success);
        assert!(task.ended_at.is_some());
        assert_eq!(task.llm_cost_usd, Decimal::new(5, 2));
        assert_eq!(task.total_cost_usd, Decimal::new(5, 2));
        assert_eq!(task.total_input_tokens, 1000);
        assert_eq!(task.total_output_tokens, 500);
        assert_eq!(task.retry_count, 0);

        // Task should be persisted in buffer.
        assert_eq!(buffer.task_count(), 1);
    }

    #[tokio::test]
    async fn test_finalize_auto_task_with_external_cost() {
        let _g = CTX_LOCK.lock().await;
        clear_dexcost_context().await;

        let mut task = create_auto_task("google_maps").await;
        let mut event = CostEvent::new(&task.task_id, EventType::ExternalCost);
        event.cost_usd = Decimal::new(1, 2); // 0.01

        let mut buffer = EventBuffer::new().unwrap();
        finalize_auto_task(&mut task, Some(&event), "success", &mut buffer);

        assert_eq!(task.external_cost_usd, Decimal::new(1, 2));
        assert_eq!(task.total_cost_usd, Decimal::new(1, 2));
    }

    #[tokio::test]
    async fn test_finalize_auto_task_with_retry_event() {
        let _g = CTX_LOCK.lock().await;
        clear_dexcost_context().await;

        let mut task = create_auto_task("openai.chat").await;
        let mut event = CostEvent::new(&task.task_id, EventType::LlmCall);
        event.cost_usd = Decimal::new(3, 2); // 0.03
        event.is_retry = true;
        event.retry_reason = Some("rate_limit".to_string());
        event.input_tokens = Some(500);
        event.output_tokens = Some(200);

        let mut buffer = EventBuffer::new().unwrap();
        finalize_auto_task(&mut task, Some(&event), "success", &mut buffer);

        assert_eq!(task.retry_count, 1);
        assert_eq!(task.retry_cost_usd, Decimal::new(3, 2));
        assert_eq!(task.llm_cost_usd, Decimal::new(3, 2));
    }

    #[tokio::test]
    async fn test_finalize_auto_task_without_event() {
        let _g = CTX_LOCK.lock().await;
        clear_dexcost_context().await;

        let mut task = create_auto_task("noop").await;

        let mut buffer = EventBuffer::new().unwrap();
        finalize_auto_task(&mut task, None, "failed", &mut buffer);

        assert_eq!(task.status, TaskStatus::Failed);
        assert!(task.ended_at.is_some());
        assert_eq!(task.total_cost_usd, Decimal::ZERO);
    }

    #[tokio::test]
    async fn test_finalize_auto_task_compute_cost() {
        let _g = CTX_LOCK.lock().await;
        clear_dexcost_context().await;

        let mut task = create_auto_task("gpu_inference").await;
        let mut event = CostEvent::new(&task.task_id, EventType::ComputeCost);
        event.cost_usd = Decimal::new(10, 2); // 0.10

        let mut buffer = EventBuffer::new().unwrap();
        finalize_auto_task(&mut task, Some(&event), "success", &mut buffer);

        assert_eq!(task.compute_cost_usd, Decimal::new(10, 2));
        assert_eq!(task.total_cost_usd, Decimal::new(10, 2));
    }
}
